File: src/main/java/org/radarcns/android/auth/OAuth2StateManager.java
/*
 * Copyright 2017 The Hyve
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.radarcns.android.auth;

import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.net.Uri;
import android.support.annotation.AnyThread;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import net.openid.appauth.AuthState;
import net.openid.appauth.AuthorizationException;
import net.openid.appauth.AuthorizationRequest;
import net.openid.appauth.AuthorizationResponse;
import net.openid.appauth.AuthorizationService;
import net.openid.appauth.AuthorizationServiceConfiguration;
import net.openid.appauth.RegistrationResponse;
import net.openid.appauth.ResponseTypeValues;
import net.openid.appauth.TokenResponse;

import org.json.JSONException;
import org.radarcns.android.RadarConfiguration;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.lang.ref.WeakReference;

import static org.radarcns.android.RadarConfiguration.OAUTH2_AUTHORIZE_URL;
import static org.radarcns.android.RadarConfiguration.OAUTH2_CLIENT_ID;
import static org.radarcns.android.RadarConfiguration.OAUTH2_TOKEN_URL;

public class OAuth2StateManager {
    private static final int OAUTH_INTENT_HANDLER_REQUEST_CODE = 8422341;

    private static WeakReference<OAuth2StateManager> INSTANCE_REF = new WeakReference<>(null);
    private static final Object INSTANCE_SYNC_OBJECT = new Object();
    private static final Logger logger = LoggerFactory.getLogger(OAuth2StateManager.class);

    private static final String STORE_NAME = "AuthState";
    private static final String KEY_STATE = "state";

    private final SharedPreferences mPrefs;
    private AuthState mCurrentAuthState;

    @AnyThread
    public static OAuth2StateManager getInstance(@NonNull Context context) {
        synchronized (INSTANCE_SYNC_OBJECT) {
            OAuth2StateManager manager = INSTANCE_REF.get();
            if (manager == null) {
                manager = new OAuth2StateManager(context.getApplicationContext());
                INSTANCE_REF = new WeakReference<>(manager);
            }
            return manager;
        }
    }

    private OAuth2StateManager(Context context) {
        mPrefs = context.getSharedPreferences(STORE_NAME, Context.MODE_PRIVATE);
        mCurrentAuthState = readState();
    }

    @AnyThread
    @NonNull
    public synchronized AuthState getAuthState() {
        return mCurrentAuthState;
    }

    @AnyThread
    public synchronized void login(@NonNull LoginActivity context, @NonNull RadarConfiguration config) {
        Uri authorizeUri = Uri.parse(config.getString(OAUTH2_AUTHORIZE_URL));
        Uri tokenUri = Uri.parse(config.getString(OAUTH2_TOKEN_URL));
        String clientId = config.getString(OAUTH2_CLIENT_ID);

        AuthorizationServiceConfiguration authConfig =
                new AuthorizationServiceConfiguration(authorizeUri, tokenUri, null);

        AuthorizationRequest.Builder authRequestBuilder =
                new AuthorizationRequest.Builder(
                        authConfig, // the authorization service configuration
                        clientId, // the client ID, typically pre-registered and static
                        ResponseTypeValues.CODE, // the response_type value: we want a code
                        Uri.parse("org.radarcns.android:/oauth2redirect/radar")); // the redirect URI to which the auth response is sent

        AuthorizationService service = new AuthorizationService(context);
        service.performAuthorizationRequest(
                authRequestBuilder.build(),
                PendingIntent.getActivity(context,
                        OAUTH_INTENT_HANDLER_REQUEST_CODE,
                        new Intent(context, context.getClass()),
                        PendingIntent.FLAG_ONE_SHOT));
    }

    @AnyThread
    public synchronized void updateAfterAuthorization(@NonNull final LoginActivity context,
                                                      @Nullable Intent intent) {
        if (intent == null) {
            return;
        }

        AuthorizationResponse resp = AuthorizationResponse.fromIntent(intent);
        AuthorizationException ex = AuthorizationException.fromIntent(intent);

        if (resp != null || ex != null) {
            mCurrentAuthState.update(resp, ex);
            writeState(mCurrentAuthState);
        }

        if (resp != null) {
            AuthorizationService service = new AuthorizationService(context);
            // authorization succeeded
            service.performTokenRequest(
                    resp.createTokenExchangeRequest(),
                    new AuthorizationService.TokenResponseCallback() {
                        @Override public void onTokenRequestCompleted(
                                TokenResponse resp, AuthorizationException ex) {
                            if (resp != null) {
                                updateAfterTokenResponse(resp, ex);
                                context.oauthSucceeded();
                            } else {
                                context.oauthFailed(ex);
                            }
                        }
                    });
        } else if (ex != null) {
            context.oauthFailed(ex);
        }
    }

    @AnyThread
    public synchronized void updateAfterTokenResponse(
            @Nullable TokenResponse response,
            @Nullable AuthorizationException ex) {
        mCurrentAuthState.update(response, ex);
        writeState(mCurrentAuthState);
    }

    @AnyThread
    public synchronized void updateAfterRegistration(
            RegistrationResponse response,
            AuthorizationException ex) {
        if (ex == null) {
            mCurrentAuthState.update(response);
            writeState(mCurrentAuthState);
        }
    }

    @AnyThread
    @NonNull
    private synchronized AuthState readState() {
        String currentState = mPrefs.getString(KEY_STATE, null);
        if (currentState == null) {
            return new AuthState();
        }

        try {
            return AuthState.jsonDeserialize(currentState);
        } catch (JSONException ex) {
            logger.warn("Failed to deserialize stored auth state - discarding", ex);
            writeState(null);
            return new AuthState();
        }
    }

    @AnyThread
    private synchronized void writeState(@Nullable AuthState state) {
        if (state != null) {
            mPrefs.edit().putString(KEY_STATE, state.jsonSerializeString()).apply();
        } else {
            mPrefs.edit().remove(KEY_STATE).apply();
        }
    }
}

File: plugins/radar-android-login-oauth2/src/main/java/org/radarbase/android/auth/oauth2/OAuth2StateManager.kt
/*
 * Copyright 2017 The Hyve
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.radarbase.android.auth.oauth2

import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.net.Uri
import android.os.Process
import androidx.activity.result.ActivityResultLauncher
import net.openid.appauth.*
import org.json.JSONException
import org.radarbase.android.RadarConfiguration
import org.radarbase.android.auth.AppAuthState
import org.radarbase.android.auth.AuthService
import org.radarbase.android.config.SingleRadarConfiguration
import org.slf4j.LoggerFactory
import androidx.core.net.toUri
import androidx.core.content.edit
import org.radarbase.android.auth.commons.AbstractRadarPortalClient
import org.radarbase.android.auth.commons.AuthType
import org.radarbase.android.auth.portal.GetSubjectParser
import org.radarbase.android.auth.sep.SEPClient
import org.radarbase.android.util.SafeHandler
import java.util.concurrent.locks.ReentrantLock

class OAuth2StateManager(private val config: RadarConfiguration, private val oauthManager: OAuth2LoginManager, context: Context) {
    private val oauthPrefs: SharedPreferences = context.getSharedPreferences(STORE_NAME, Context.MODE_PRIVATE)
    private var currentAuthState: AuthState
    private val oAuthService: AuthorizationService = AuthorizationService(context)
    private val handler: SafeHandler = SafeHandler.getInstance("OAuth2StateManager-Handler", Process.THREAD_PRIORITY_BACKGROUND)
    private val refreshLock = ReentrantLock()

    init {
        currentAuthState = readState()
        handler.start()
    }

    fun login(
        config: SingleRadarConfiguration,
        activityResultLauncher: ActivityResultLauncher<Intent>,
    ) {
        if (refreshLock.tryLock()) {
            try {
                val (authorizeUri: Uri, tokenUri: Uri, redirectUri: Uri, clientId: String) = getOAuthConfig(config)
                logger.debug("Performing oAuth request at authorization url: {}", authorizeUri)

                val authConfig = AuthorizationServiceConfiguration(authorizeUri, tokenUri)
                val additionalParams = mapOf(
                    "audience" to DEFAULT_APP_RESOURCES
                )

                val authRequestBuilder = AuthorizationRequest.Builder(
                    authConfig,
                    clientId,
                    ResponseTypeValues.CODE,
                    redirectUri,
                ).apply {
                    setScopes(DEFAULT_OAUTH_SCOPES)
                    setAdditionalParameters(additionalParams)
                }

                logger.info("Performing auth request")
                val authIntent = oAuthService.getAuthorizationRequestIntent(authRequestBuilder.build())
                activityResultLauncher.launch(authIntent)
            } finally {
                refreshLock.unlock()
            }
        }
    }

    private fun getOAuthConfig(config: SingleRadarConfiguration): OAuthConfig {
        try {
            val authorizeUri = config.getString(RadarConfiguration.OAUTH2_AUTHORIZE_URL).toUri()
            val tokenUri = config.getString(RadarConfiguration.OAUTH2_TOKEN_URL).toUri()
            val redirectUri = config.getString(RadarConfiguration.OAUTH2_REDIRECT_URL, APP_REDIRECT_URI).toUri()
            val clientId = config.getString(RadarConfiguration.OAUTH2_CLIENT_ID)
            config.getString(RadarConfiguration.OAUTH2_CLIENT_SECRET)

            return OAuthConfig(authorizeUri, tokenUri, redirectUri, clientId)
        } catch (e: IllegalArgumentException) {
            throw MissingConfigurationException("Missing firebase remote configuration for either oauth endpoints, redirect uri or client credentials.", e)
        }
    }

    private fun logIntent(intent: Intent?) {
        val data = intent?.data
        val action = intent?.action
        val flags = intent?.flags
        val extrasKeys = intent?.extras?.keySet()?.joinToString(", ") ?: "null"
        logger.warn("(TestOauthDebug) AuthRequestDebug: action=$action, data=$data, flags=$flags, extras=$extrasKeys")
        if (intent?.extras != null) {
            intent.extras!!.keySet().forEach { k ->
                logger.warn("(TestOauthDebug) AuthRequestDebug: extra $k => ${intent.extras?.get(k)}")
            }
        }
    }

        fun updateAfterAuthorization(
        authService: AuthService,
        intent: Intent?,
        binder: AuthService.AuthServiceBinder,
        client: AbstractRadarPortalClient?
    ) {
        if (refreshLock.tryLock()) {
            try {
                if (intent == null) {
                    logger.info("Intent is null ")
                    return
                }
                logIntent(intent)
                val resp = AuthorizationResponse.fromIntent(intent)
                val ex = AuthorizationException.fromIntent(intent)

                if (resp == null && ex == null) {
                    logger.warn("Both response and exception are null for authorization redirect intent")
                }

                if (resp != null || ex != null) {
                    currentAuthState.update(resp, ex)
                    writeState(currentAuthState)
                }

                val clientAuth = clientSecretBasic()

                if (resp != null) {
                    binder.applyState {
                        oAuthService.performTokenRequest(
                            resp.createTokenExchangeRequest(),
                            clientAuth,
                            processTokenResponse(authService, this@applyState, client)
                        )
                    }
                } else if (ex != null) {
                    authService.loginFailed(null, ex)
                }
            } finally {
                refreshLock.unlock()
            }
        }
    }

    fun refresh(
        context: AuthService,
        authState: AppAuthState,
        refreshToken: String?,
        client: AbstractRadarPortalClient?
    ) {
        // refreshToken does not originate from the current auth state.
        if (refreshLock.tryLock()) {
            try {
                val clientAuth = clientSecretBasic()
                if (refreshToken != null && refreshToken != currentAuthState.refreshToken) {
                    try {
                        val json = currentAuthState.jsonSerialize()
                        json.put("refreshToken", refreshToken)
                        currentAuthState = AuthState.jsonDeserialize(json)
                    } catch (ex: JSONException) {
                        logger.error("Failed to update refresh token", ex)
                    }
                }
                // authorization succeeded
                oAuthService.performTokenRequest(
                    currentAuthState.createTokenRefreshRequest(),
                    clientAuth,
                    processTokenResponse(context, authState, client),
                )
            } finally {
                refreshLock.unlock()
            }
        }
    }

    private fun clientSecretBasic(): ClientSecretBasic {
        return config.latestConfig.getString(RadarConfiguration.OAUTH2_CLIENT_SECRET)
            .let { clientSecret ->
                ClientSecretBasic(clientSecret)
            }
    }

    private fun processTokenResponse(
        context: AuthService,
        authState: AppAuthState,
        client: AbstractRadarPortalClient?,
    ) = AuthorizationService.TokenResponseCallback { resp, ex ->
        resp ?: return@TokenResponseCallback context.loginFailed(null, ex)
        updateAfterTokenResponse(resp, ex)
        val accessTokenParser = OAuthAccessTokenParser()
        val updatedAuth = accessTokenParser.parse(authState, currentAuthState)

        logger.warn("(TestOauthDebug) Retrieved access token successfully, now getting subjects on thread: {}", Thread.currentThread().name)
        config.updateWithAuthState(context, updatedAuth)
        val getSubjectParser = GetSubjectParser(updatedAuth, AuthType.OAUTH2)
        if (client is SEPClient) {
            handler.compute {
                client.getSubject(updatedAuth, getSubjectParser)
            }.also { latestAuth ->
                oauthManager.loginSucceeded(null, latestAuth)
            }
        } else {
            logger.error("Can't process the get subject request, client is not the instance of SEP Client")
        }

    }

    @Synchronized
    fun updateAfterTokenResponse(
        response: TokenResponse?,
        ex: AuthorizationException?
    ) {
        currentAuthState.update(response, ex)
        writeState(currentAuthState)
    }

    @Synchronized
    private fun readState(): AuthState {
        val currentState = oauthPrefs.getString(KEY_STATE, null) ?: return AuthState()

        return try {
            AuthState.jsonDeserialize(currentState)
        } catch (ex: JSONException) {
            logger.warn("Failed to deserialize stored auth state - discarding", ex)
            writeState(null)
            AuthState()
        }

    }

    @Synchronized
    private fun writeState(state: AuthState?) {
        oauthPrefs.edit {
            if (state != null) {
                putString(KEY_STATE, state.jsonSerializeString())
            } else {
                remove(KEY_STATE)
            }
        }
    }

    fun stop() {
        oAuthService.dispose()
        handler.stop()
    }

    companion object {
        private val logger = LoggerFactory.getLogger(OAuth2StateManager::class.java)

        private const val STORE_NAME = "AuthState"
        private const val KEY_STATE = "state"
        private const val APP_REDIRECT_URI = "org.radarbase.prmt://login"
        private const val DEFAULT_OAUTH_SCOPES = "SUBJECT.READ SUBJECT.UPDATE MEASUREMENT.CREATE offline_access"
        private const val DEFAULT_APP_RESOURCES = "res_gateway res_ManagementPortal res_appconfig"

        class MissingConfigurationException(message: String) : RuntimeException(message) {
            constructor(message: String, cause: Throwable) : this(message) {
                initCause(cause)
            }
        }

        private data class OAuthConfig(
            val authorizeUri: Uri,
            val tokenUri: Uri,
            val redirectUri: Uri,
            val clientId: String,
        )
    }
}
